internal/guest/storage/scsi/scsi.go

```diff
@@ -357,3 +415,60 @@ func UnplugDevice(ctx context.Context, controller, lun uint8) (err error) {
 	}
 	return nil
 }
+
+// runBlkidCommand execs a call to `blkid`. This function is broken out
+// so that we can stub it for unit testing of this package.
+func runBlkidCommand(args []string) (string, error) {
+	cmd := exec.Command("blkid", args...)
+	raw, err := cmd.Output()
+	return string(raw), err
+}
+
+// getDeviceFsType queries to find the filesystem on a dev device.
+// The mount(8) man page states that it uses `blkid` to determine
+// what filesystem a given device is using, so we do the same here.
+//
+// If a device is present but has no filesystem configured, which would be the
+// case for empty scratch disks for example, they will not show up in the
+// response from `blkid`, and `blkid` will return an error. Since `blkid`
+// returns the same error both for if a device is not present or a device
+// does not have a filesystem configured on it, getDeviceFsType should only
+// be called after we've validated that the device exists.
+//
+// `devicePath` is the /dev path to the target device.
+func getDeviceFsType(devicePath string) (string, error) {
+	rawOutput, err := _runBlkidCommand([]string{devicePath})
+	if err != nil {
+		// According to the blkid man page, blkid will return 2 if the specified token
+		// was not found, or no (specified) devices could be identified.
+		// blkid will return 4 for "usage or other errors".
+		if exiterr, ok := err.(*exec.ExitError); ok && exiterr.ExitCode() == 2 {
+			// since we have already validated that the device exists under /dev,
+			// we know at this point that this device has no filesystem configured
+			// for it, so return an empty string and no error.
+			return "", nil
+		}
+		return "", err
+	}
+
+	// blkid returns something that looks like the following:
+	//
+	// 		/dev/sda: TYPE="ext4"
+	//
+	// 			OR
+	//
+	// 		/dev/sdb: UUID="3255683f-53a2-4fdf-91cf-b4c1041e2a62" TYPE="ext4"
+	//
+	// parse the raw output to get device type
+	fields := strings.Split(string(rawOutput), " ")
+	if len(fields) < 2 {
+		return "", fmt.Errorf("expected to get at least two fields in response for querying %v with blkid: %w", devicePath, err)
+	}
+	for _, item := range fields {
+		if strings.HasPrefix(item, "TYPE=") {
+			filesystemType := strings.Trim(strings.TrimPrefix(item, "TYPE="), "\"")
+			return filesystemType, nil
+		}
+	}
```

Comment:
Consider using a regex here instead?

Reply:
Updated to use regex. I'm not super familiar/comfortable with regex so please take another look to make sure it looks good. I did test the regex string to make sure it compiles and returns what I expect.

Reply:
This is better done with submatches: https://go.dev/play/p/TxNQMuqjKXM

The `()` in the regex surrounds a submatch that can be extracted separately.

Reply:
Removed the regex entirely in place of reading the device's superblock. Resolving this.